asset-transfer-ledger-queries/chaincode-javascript/lib/asset_transfer_ledger_chaincode.js

```diff
@@ -0,0 +1,425 @@
+
+'use strict';
+
+const { Contract } = require('fabric-contract-api');
+
+class Chaincode extends Contract{
+
+  // CreateAsset - create a new asset, store into chaincode state
+  async createAsset(ctx, assetID, color, owner, size, appraisedValue) {
+    const exists = await this.assetExists(ctx, assetID)
+    if (exists) {
+      throw new Error('asset exists')
+    }
+
+    // ==== Create asset object and marshal to JSON ====
+    let asset = {};
+    asset.docType = 'asset';
+    asset.ID = assetID;
+    asset.color = color;
+    asset.size = size;
+    asset.owner = owner;
+    asset.appraisedValue = appraisedValue;
+
+    // === Save asset to state ===
+    await ctx.stub.putState(assetID, Buffer.from(JSON.stringify(asset)));
+    let indexName = 'color~name'
+    let colorNameIndexKey = await ctx.stub.createCompositeKey(indexName, [asset.color, asset.ID]);
+
+    //  Save index entry to state. Only the key name is needed, no need to store a duplicate copy of the marble.
+    //  Note - passing a 'nil' value will effectively delete the key from state, therefore we pass null character as value
+    await ctx.stub.putState(colorNameIndexKey, Buffer.from('\u0000'));
+  }
+
+  // readAsset returns the asset stored in the world state with given id.
+  async readAsset(ctx, id) {
+    const assetJSON = await ctx.stub.getState(id); // get the asset from chaincode state
+    if (!assetJSON || assetJSON.length === 0) {
+        throw new Error(`The asset ${id} does not exist`);
+    }
+
+    return assetJSON.toString();
+  }
+
+  // delete - remove a asset key/value pair from state
+  async deleteAsset(ctx, id) {
+    if (!id) {
+      throw new Error('asset name must not be empty');
+    }
+
+    var exists = await this.assetExists(ctx, id)
+    if (!exists) {
+      throw new Error('')
+    }
+
+    // to maintain the color~name index, we need to read the asset first and get its color
+    let valAsbytes = await ctx.stub.getState(id); // get the asset from chaincode state
+    let jsonResp = {};
+    if (!valAsbytes) {
+      jsonResp.error = 'asset does not exist: ' + name;
+      throw new Error(jsonResp);
+    }
+    let assetJSON = {};
+    try {
+      assetJSON = JSON.parse(valAsbytes.toString());
+    } catch (err) {
+      jsonResp = {};
+      jsonResp.error = 'Failed to decode JSON of: ' + id;
+      throw new Error(jsonResp);
+    }
+    await ctx.stub.deleteState(id); //remove the asset from chaincode state
+
+    // delete the index
+    let indexName = 'color~name';
+    let colorNameIndexKey = ctx.stub.createCompositeKey(indexName, [assetJSON.color, assetJSON.ID]);
+    if (!colorNameIndexKey) {
+      throw new Error(' Failed to create the createCompositeKey');
```

Comment:
nit: there seems to be a mix-match of using uppercase and lowercase for error messages. 
Also understand this might be come from `asset-transfer-basci` chaincode-js, that can be fixed with a separate PR.